Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -299,9 +299,11 @@ public final class HTTPServerPipelineHandler: ChannelDuplexHandler, RemovableCha
         debugOnly {
             let res = self.unwrapOutboundIn(data)
             switch res {
-            case .head:
+            case .head(let head):
                 assert(self.nextExpectedOutboundMessage == .head)
-                self.nextExpectedOutboundMessage = .bodyOrEnd
+                if !head.status.isInformational || head.status == .switchingProtocols {
```

Comment:
🤔

Reply:
hey @weissi, I was looking into this again. I'm not sure I understand the scope of this change anymore. Isn't the idea to be able to send a `.head(non-100)` after a previous `.head(100)`? What is the purpose of a `.end` in between? isn't `.end` a noop in this case, i.e. nothing will be written to the channel?

Reply:
If I introduce the intermediary `.end` in between `1xx` responses, then the state machine in `HTTPServerPipelineHandler` is not happy, because it thinks that the server already completed its response. If I omit the `.end` in between, everything seems to work™.

Reply:
@t089 in that case we need to adapt the state machine to correctly recognise that case.
 Adding @Lukasa for input. In think it's important that we retain the guarantee that `.head` and `.end` are balanced. On the other hand, having "two responses" (the `100 continue` & then the actual one) is also breaking an invariant that previously held...

Maybe we should just handle the `expect: 100-continue` automatically?

Reply:
I don't think we can make this change without _someone_ being wrong about what they expected. In that instance, we may want to be led by `http_decoder`. If it receives a `100 Continue` followed by the status code, what events does it generate? We should arrange for the two sides to match.

Reply:
@Lukasa well, we could add automatic handling in a handler that we add through `configureHTTPServerHandlers`. Then at least if you use that you won't notice...

Reply:
I'm not sure I follow what you're proposing there.

Reply:
@Lukasa Well, I think it's important that our _users_ still receive and send `.head` & `.end` balanced as well has having one request `.head` for one response `.head`. If that's no longer true, a lot of stuff will break.

That leaves us I think with two options:

1. make a channel handler that just sends the `100 continue` automatically.
2. use an outbound user event to send it.

On the client side, we're in a pretty similar situation I guess.


I think I just convinced myself that user events are a good option for this. the whole 100 continue is information after all...

Reply:
For user events, that's fine, but are we doing that for the entire 1xx range? It's a huge change from the current behaviour.

Reply:
@Lukasa Well, right now, we expect 1 request / 1 response. That currently means we don't support informal responses, right?

Reply:
At least 101 clearly works.

Reply:
@Lukasa right, because it's switching protocols...

Reply:
it retains the 1 request / 1 response model for the HTTP part which is probably why we could punt on this for so long :P.

Reply:
Sure, but that means that we can't just use a big heavy "move 1XXes to user events" without causing breakage.

Reply:
We can do "move everything but 101", but that creates a potentially awkward special case. I'm not _desperately_ opposed to that change though, I think it's somewhat defensible.

Reply:
Or we can just fix our stuff and say "the 1 request 1 response model was never in line with the HTTP specification and anyone who assumed it was was in error".

Reply:
@Lukasa I don't think I agree. For 101 it happens to work as we currently do it. I don't think we need to precondition when you send a 1xx but I don't think we should add support for first sending 1xx, then sending Yxx (where Y != 1).

Reply:
That keeps everything working as is and for everything else we can add a nice user event.

Reply:
User events are tricky because naive handlers won't buffer them in line with data.

Reply:
That's true, however, they would only exist between `HTTPEncoder` and the issuing handler which almost certainly only has our handlers. So we can make our handlers get that right.

Reply:
Sure, but if it's only our handlers in between `HTTPEncoder` and the issuing handler for 1XX then we can just make them get multiple `.head`/`.end` right. :wink:

Reply:
@Lukasa I don't think so because I think it's much more conceivable that someone relies on the 1 req / 1 res property through `channelRead` and `write` than it is that they're trying to buffer.
The reason I doubt they'll buffer is that there's pretty much no reason because they're (hopefully) using the pipeline handler.

Reply:
> That's true, however, they would only exist between HTTPEncoder and the issuing handler which almost certainly only has our handlers.

> I don't think so because I think it's much more conceivable that someone relies on the 1 req / 1 res property through channelRead and write than it is that they're trying to buffer.

I think these two statements are in tension. My understanding of what you're saying is that you're proposing that we adopt a behaviour based on a user you think probably doesn't exist, but if they did exist, you think they probably don't buffer, but probably do assume one `.head`/`.end` out per `.head`/`.end` in.

It seems to me that the more important factor is the first than the second. If we think that no such user exists, then I'd propose we change the model to match the semantic of the protocol. This is what HTTP/2 does: 1XX header blocks are not hidden from the user in user events, but instead you are required to know that you might see multiple HEADERS frames. It's also what we do for trailers in both HTTP/2 and HTTP/1: both are part of the regular data stream, and you are required to know about them.

I think the bug here that NIO incorrectly made assumptions about the HTTP/1 protocol that are not valid. It was wrong to do so, and should become right. In most cases, users will not see this bug because they will not operate on this level: servers will only see 1XX responses if they themselves choose to send them, clients will mostly use async-http-client and not operate at this layer. Users that _do_ operate at this layer are, in my view, obligated to understand how the protocol works and so to deal with the reality of 1XX headers. This also affects things like proxies which are required to pass 1XX headers on, requiring that any proxy implementer know about the user event and correctly sequence it through their pipeline.

If we think that users _do_ exist at this level, the user event puts the lie to their mental model anyway. They'll now be missing important protocol level events. If one of the things those handlers did was attempt to log, for example, the log is now missing entire parts of the HTTP dialog.

For my part the particularly negative part of the user event is that we're planning to exempt 101 from it. This indicates that we don't really _mean_ the user event. We're proposing to put this in the user event not because it's a good solution but because we're worried about breaking users. Users that, as I understand it, you don't think exist. I don't think those users exist either, and even if they did my theory is they'll be broken by the user event but _more subtly_. So why not just arrange to do this the way we'd (almost) do it from the start?

- - - 

Digression: I say "almost" in the above paragraph because clearly the way we'd have done this from the start if we were _smart_ is to have a fourth case in the `HTTP{Client,Server}ResponsePart` enumerations for 1XX header blocks. Can't do that now though, whoops!

Reply:
@Lukasa 100% agree that the right solution would be to add a fourth case but we can't before NIO 3. Just so we won't forget, I filed it as https://github.com/apple/swift-nio/issues/1525 .

Your argument for the server side is very convincing. They won't break because they won't emit 1xx (!= 101)s today.

For the client side, I'm less convinced but I think I've got an idea too: Why don't we add a parameter whether we support informal responses to the parser constructor? We default it to `false` which means we're dropping them on the floor. Then clients can opt in and everything's good?

Reply:
I'm happy with that solution. It's even in line with RFC 7231 (which allows dropping unexpected 1XX responses).

We can even try to detect bad situations in debug mode (e.g. configuring "I don't support 1XX responses" and then sending `Expect: 100-continue`).

Reply:
Oh nice! Good job RFC 7231 :). Let's do it then.

We need some documentation on what exactly the users can now expect but I think this really fixes the problem and breaks nothing.

@t089 are we making some sense?

Reply:
Hey @weissi and @Lukasa, sorry for my silence. I'd love to continue working on this and I think I also understand most of your discussion. Unfortunately, due to 😷 situation and 👶, I had not enough time, I hope I can soon tackle this again and will follow up with questions.